tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Grpc/MetadataHeadersCollection.cs

```diff
@@ -0,0 +1,59 @@
+// <copyright file="MetadataHeadersCollection.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+#nullable enable
+
+using System.Collections.Generic;
+using System.Linq;
+using Datadog.Trace.DuckTyping;
+using Datadog.Trace.Headers;
+using Datadog.Trace.Util;
+
+namespace Datadog.Trace.ClrProfiler.AutoInstrumentation.Grpc
+{
+    internal readonly struct MetadataHeadersCollection : IHeadersCollection
+    {
+        private readonly IMetadata _headers;
+
+        public MetadataHeadersCollection(IMetadata? headers)
+        {
+            if (headers is null)
+            {
+                ThrowHelper.ThrowArgumentNullException(nameof(headers));
+            }
+
+            _headers = headers;
+        }
+
+        public IEnumerable<string> GetValues(string name)
+        {
+            return from object? header in _headers.GetAll(name)
+                   where header is not null
+                   select header.DuckCast<MetadataEntryStruct>() into entry
```

Comment:
I haven't seen this `select ... into ...` LINQ syntax before. It looks like this line is equivalent to  `let entry = header.DuckCast<MetadataEntryStruct>()`, except that `header` goes out of scope. Is that correct?

Reply:
Looking at the generated IL, your code looks simpler and avoids allocations for the `{ header, header.DuckCast<MetadataEntryStruct>() }` anonymous type in my example.

TIL!